addons/nuxeo-retention/nuxeo-retention-core/src/main/java/org/nuxeo/retention/actions/EvalInputEventBasedRuleAction.java

```diff
@@ -0,0 +1,110 @@
+import org.nuxeo.ecm.core.bulk.action.computation.AbstractBulkComputation;
+import org.nuxeo.ecm.core.bulk.message.BulkCommand;
+import org.nuxeo.ecm.platform.audit.service.NXAuditEventsService;
+import org.nuxeo.lib.stream.computation.Topology;
+import org.nuxeo.retention.RetentionConstants;
+import org.nuxeo.retention.adapters.Record;
+import org.nuxeo.retention.adapters.RetentionRule;
+import org.nuxeo.retention.service.RetentionManager;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.stream.StreamProcessorTopology;
+
+/**
+ * Bulk action to evaluate expression on record documents with an attached event-based retention rule. Depending on the
+ * expression evaluation outcome, a determinate retention period is computed and set on the record document.
+ *
+ * @since 11.1
+ */
+public class EvalInputEventBasedRuleAction implements StreamProcessorTopology {
+
+    public static final String ACTION_NAME = "evalInputEventBasedRule";
+
+    public static final String PARAM_EVENT = "event";
+
+    @Override
+    public Topology getTopology(Map<String, String> options) {
+        return Topology.builder()
+                       .addComputation(EvalEventBasedRuleComputation::new,
+                               Arrays.asList(INPUT_1 + ":" + ACTION_NAME, OUTPUT_1 + ":" + STATUS_STREAM))
+                       .build();
+    }
+
+    public static class EvalEventBasedRuleComputation extends AbstractBulkComputation {
+
+        static final Logger log = LogManager.getLogger(EvalEventBasedRuleComputation.class);
+
+        protected boolean disableAudit;
+
+        protected RetentionManager retentionManager;
+
+        public EvalEventBasedRuleComputation() {
+            super(ACTION_NAME);
+        }
+
+        @Override
+        public void startBucket(String bucketKey) {
+            BulkCommand command = getCurrentCommand();
+            Serializable auditParam = command.getParam(NXAuditEventsService.DISABLE_AUDIT_LOGGER);
+            disableAudit = auditParam != null && Boolean.parseBoolean(auditParam.toString());
+            retentionManager = Framework.getService(RetentionManager.class);
+        }
+
+        @Override
+        protected void compute(CoreSession session, List<String> ids, Map<String, Serializable> properties) {
+            for (DocumentModel recordDoc : loadDocuments(session, ids)) {
+                if (!recordDoc.hasFacet(RetentionConstants.RECORD_FACET)) {
+                    log.debug("Document {} is not a record, ignoring ...", recordDoc::getPathAsString);
+                    continue;
+                }
+
+                Record record = recordDoc.getAdapter(Record.class);
+                if (!record.isRetainUntilInderterminate()) {
+                    log.debug("Record {} has already a determinate retention date {}, ignoring ...",
+                            recordDoc::getPathAsString,
+                            () -> RetentionConstants.DEFAULT_DATE_FORMAT.format(recordDoc.getRetainUntil()));
```

Comment:
you can use a static import for more readability. wdty ;)

Reply:
Not sure, `DEFAULT_DATE_FORMAT` by itself wouldn't give much context about which default. I'll probably change this to a standard ISO formatter later anyway.